File: skills/light-result-analysis/scripts/r_analysis_crosscheck.R
#!/usr/bin/env Rscript

# Base-R cross-check for one two-group comparison.  No optional package is
# required; the output is a one-row CSV so Python and R can be compared exactly.

parse_args <- function(args) {
  out <- list()
  i <- 1
  while (i <= length(args)) {
    key <- args[[i]]
    if (key == "--selftest") {
      out$selftest <- TRUE
      i <- i + 1
    } else {
      if (i == length(args)) stop(paste("missing value for", key))
      out[[sub("^--", "", key)]] <- args[[i + 1]]
      i <- i + 2
    }
  }
  out
}

analyze_two_groups <- function(input, group, metric, paired_by = NULL) {
  d <- read.csv(input, check.names = FALSE, stringsAsFactors = FALSE)
  needed <- c(group, metric)
  if (!is.null(paired_by)) needed <- c(needed, paired_by)
  missing <- setdiff(needed, names(d))
  if (length(missing) > 0) stop(paste("missing columns:", paste(missing, collapse = ",")))

  d <- d[complete.cases(d[, needed, drop = FALSE]), needed, drop = FALSE]
  groups <- sort(unique(as.character(d[[group]])))
  if (length(groups) != 2) stop("exactly two groups are required")

  if (!is.null(paired_by)) {
    a <- d[d[[group]] == groups[[1]], c(paired_by, metric), drop = FALSE]
    b <- d[d[[group]] == groups[[2]], c(paired_by, metric), drop = FALSE]
    if (anyDuplicated(a[[paired_by]]) || anyDuplicated(b[[paired_by]])) {
      stop("paired key must be unique within each group")
    }
    names(a)[2] <- "value_a"
    names(b)[2] <- "value_b"
    m <- merge(a, b, by = paired_by, all = FALSE)
    if (nrow(m) < 2) stop("fewer than two complete pairs")
    test <- t.test(m$value_a, m$value_b, paired = TRUE, conf.level = 0.95)
    diff <- m$value_a - m$value_b
    effect <- if (sd(diff) == 0) NA_real_ else mean(diff) / sd(diff)
    n <- nrow(m)
    design <- "paired"
  } else {
    x <- d[d[[group]] == groups[[1]], metric]
    y <- d[d[[group]] == groups[[2]], metric]
    if (length(x) < 2 || length(y) < 2) stop("fewer than two rows per group")
    test <- t.test(x, y, paired = FALSE, var.equal = FALSE, conf.level = 0.95)
    pooled <- sqrt(((length(x) - 1) * var(x) + (length(y) - 1) * var(y)) /
                   (length(x) + length(y) - 2))
    effect <- if (pooled == 0) NA_real_ else (mean(x) - mean(y)) / pooled
    n <- min(length(x), length(y))
    design <- "independent_welch"
  }

  data.frame(
    group1 = groups[[1]],
    group2 = groups[[2]],
    metric = metric,
    design = design,
    n = n,
    mean_diff = unname(test$estimate[[1]] - if (length(test$estimate) > 1) test$estimate[[2]] else 0),
    statistic = unname(test$statistic),
    df = unname(test$parameter),
    p = test$p.value,
    ci_low = test$conf.int[[1]],
    ci_high = test$conf.int[[2]],
    standardized_effect = effect,
    stringsAsFactors = FALSE
  )
}

selftest <- function() {
  path <- tempfile(fileext = ".csv")
  d <- data.frame(
    method = rep(c("a", "b"), each = 6),
    seed = rep(1:6, 2),
    score = c(.80, .82, .81, .83, .84, .85, .76, .77, .75, .785, .785, .80)
  )
  write.csv(d, path, row.names = FALSE)
  r <- analyze_two_groups(path, "method", "score", "seed")
  stopifnot(r$design == "paired", r$n == 6, abs(r$mean_diff - 0.05) < 1e-12)
  independent <- analyze_two_groups(path, "method", "score")
  stopifnot(
    independent$design == "independent_welch",
    independent$n == 6,
    abs(independent$mean_diff - 0.05) < 1e-12,
    is.finite(independent$p)
  )
  unlink(path)
  cat("[selftest] PASS r_analysis_crosscheck (base R paired + independent Welch t-tests)\n")
}

args <- parse_args(commandArgs(trailingOnly = TRUE))
if (isTRUE(args$selftest)) {
  selftest()
  quit(status = 0)
}
required <- c("input", "group", "metric", "out")
missing <- required[!vapply(required, function(x) !is.null(args[[x]]), logical(1))]
if (length(missing) > 0) stop(paste("missing args:", paste(missing, collapse = ",")))
paired_by <- args[["paired-by"]]
result <- analyze_two_groups(args$input, args$group, args$metric, paired_by)
write.csv(result, args$out, row.names = FALSE, quote = TRUE)
cat(paste("wrote", args$out, "\n"))

File: skills/light-result-analysis/scripts/r_analysis_crosscheck.py
#!/usr/bin/env python3
"""Locate Rscript honestly and run the base-R result cross-check."""
from __future__ import annotations

import argparse
import csv
import os
import pathlib
import shutil
import subprocess
import sys
import tempfile


HERE = pathlib.Path(__file__).resolve().parent
R_SCRIPT = HERE / "r_analysis_crosscheck.R"


def find_rscript() -> str | None:
    explicit = os.environ.get("RSCRIPT")
    if explicit and pathlib.Path(explicit).is_file():
        return explicit
    found = shutil.which("Rscript")
    if found:
        return found
    if os.name == "nt":
        root = pathlib.Path(os.environ.get("ProgramFiles", r"C:\Program Files")) / "R"
        candidates = sorted(root.glob("R-*/bin/Rscript.exe"), reverse=True)
        if candidates:
            return str(candidates[0])
    return None


def run_crosscheck(input_path: str, group: str, metric: str, out: str,
                   paired_by: str | None = None) -> subprocess.CompletedProcess:
    rscript = find_rscript()
    if not rscript:
        raise RuntimeError(
            "Rscript not found in PATH, RSCRIPT, or Windows Program Files; R cross-check not run")
    cmd = [rscript, str(R_SCRIPT), "--input", input_path, "--group", group,
           "--metric", metric, "--out", out]
    if paired_by:
        cmd += ["--paired-by", paired_by]
    return subprocess.run(cmd, check=False, text=True, capture_output=True, encoding="utf-8")


def _selftest() -> int:
    rscript = find_rscript()
    if not rscript:
        print("[selftest] SKIP r_analysis_crosscheck: Rscript unavailable (honest degradation)")
        return 0
    direct = subprocess.run(
        [rscript, str(R_SCRIPT), "--selftest"],
        check=False, text=True, capture_output=True, encoding="utf-8")
    if direct.returncode:
        print(direct.stdout)
        print(direct.stderr, file=sys.stderr)
        return direct.returncode

    with tempfile.TemporaryDirectory() as td:
        inp = pathlib.Path(td) / "results.csv"
        out = pathlib.Path(td) / "r.csv"
        with inp.open("w", newline="", encoding="utf-8") as fh:
            writer = csv.writer(fh)
            writer.writerow(["method", "seed", "score"])
            a_values = (.80, .82, .81, .83, .84, .85)
            differences = (.04, .05, .06, .045, .055, .05)
            for seed, value in enumerate(a_values, 1):
                writer.writerow(["a", seed, value])
            for seed, (value, diff) in enumerate(zip(a_values, differences), 1):
                writer.writerow(["b", seed, value - diff])
        proc = run_crosscheck(str(inp), "method", "score", str(out), "seed")
        if proc.returncode:
            print(proc.stdout)
            print(proc.stderr, file=sys.stderr)
            return proc.returncode
        with out.open(encoding="utf-8") as fh:
            row = next(csv.DictReader(fh))
        assert row["design"] == "paired" and int(row["n"]) == 6
        assert abs(float(row["mean_diff"]) - .05) < 1e-10
    print(f"[selftest] PASS r_analysis_crosscheck via {rscript}")
    return 0


def main() -> int:
    ap = argparse.ArgumentParser(description="Run a base-R two-group cross-check")
    ap.add_argument("--input")
    ap.add_argument("--group")
    ap.add_argument("--metric")
    ap.add_argument("--paired-by")
    ap.add_argument("--out")
    ap.add_argument("--selftest", action="store_true")
    args = ap.parse_args()
    if args.selftest:
        return _selftest()
    required = (args.input, args.group, args.metric, args.out)
    if not all(required):
        ap.error("--input --group --metric --out are required")
    try:
        proc = run_crosscheck(args.input, args.group, args.metric, args.out, args.paired_by)
    except RuntimeError as exc:
        print(f"[R unavailable] {exc}", file=sys.stderr)
        return 2
    if proc.stdout:
        print(proc.stdout, end="")
    if proc.stderr:
        print(proc.stderr, end="", file=sys.stderr)
    return proc.returncode


if __name__ == "__main__":
    raise SystemExit(main())
